File: wrappers/csharp/RankCache.cs
using System;
using System.IO;
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;

namespace TurboToken
{
    /// <summary>
    /// Cache for downloaded rank files.
    /// </summary>
    public static class RankCache
    {
        private static readonly HttpClient HttpClient = new HttpClient();

        /// <summary>
        /// Returns the cache directory for rank files.
        /// </summary>
        public static string CacheDir
        {
            get
            {
                var xdgCache = Environment.GetEnvironmentVariable("XDG_CACHE_HOME");
                var baseDir = !string.IsNullOrEmpty(xdgCache)
                    ? xdgCache
                    : Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".cache");
                return Path.Combine(baseDir, "turbotoken");
            }
        }

        /// <summary>
        /// Ensure a rank file is downloaded and cached. Returns the file path.
        /// </summary>
        public static async Task<string> EnsureRankFileAsync(string name, CancellationToken ct = default)
        {
            var spec = Registry.GetEncodingSpec(name);
            var uri = new Uri(spec.RankFileUrl);
            var fileName = Path.GetFileName(uri.AbsolutePath);
            var localPath = Path.Combine(CacheDir, fileName);

            if (File.Exists(localPath))
                return localPath;

            Directory.CreateDirectory(CacheDir);

            var data = await HttpClient.GetByteArrayAsync(spec.RankFileUrl).ConfigureAwait(false);
            var tempPath = localPath + ".tmp";
            File.WriteAllBytes(tempPath, data);
            File.Move(tempPath, localPath);

            return localPath;
        }

        /// <summary>
        /// Read a rank file's contents, downloading if necessary.
        /// </summary>
        public static async Task<byte[]> ReadRankFileAsync(string name, CancellationToken ct = default)
        {
            var filePath = await EnsureRankFileAsync(name, ct).ConfigureAwait(false);
            return File.ReadAllBytes(filePath);
        }
    }
}

File: wrappers/fsharp/RankCache.fs
module TurboToken.RankCache

open System
open System.IO
open System.Net.Http

let private httpClient = new HttpClient()

let cacheDir () : string =
    let xdgCache = Environment.GetEnvironmentVariable("XDG_CACHE_HOME")
    let baseDir =
        if not (String.IsNullOrEmpty xdgCache) then xdgCache
        else Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".cache")
    Path.Combine(baseDir, "turbotoken")

let ensureRankFile (name: string) : Async<Result<string, string>> =
    async {
        match Registry.getEncodingSpec name with
        | Error e -> return Error e
        | Ok spec ->
            let uri = Uri(spec.RankFileUrl)
            let fileName = Path.GetFileName(uri.AbsolutePath)
            let localPath = Path.Combine(cacheDir (), fileName)

            if File.Exists localPath then
                return Ok localPath
            else
                Directory.CreateDirectory(cacheDir ()) |> ignore
                let! data = httpClient.GetByteArrayAsync(spec.RankFileUrl) |> Async.AwaitTask
                let tempPath = localPath + ".tmp"
                File.WriteAllBytes(tempPath, data)
                File.Move(tempPath, localPath)
                return Ok localPath
    }

let readRankFile (name: string) : Async<Result<byte[], string>> =
    async {
        let! pathResult = ensureRankFile name
        match pathResult with
        | Error e -> return Error e
        | Ok filePath -> return Ok (File.ReadAllBytes filePath)
    }
